### PR TITLE
SoundPlayer: Correct wrong FFT visualization

### PR DESCRIPTION
**Pull request by standardexe**
Some small things:

- The time samples were `abs()`ed before being transformed, which just adds a DC offset and distorts the spectrum.
- Only the real part of the frequency bins were visualized, therefore encoding phase information in the spectrum. But if you play a single sine wave sound, you expect only a single bar to show in the spectrum, not a bunch.
- No window was applied to the time samples before transformation, so the spectrum had spectral leakage.
- Normalization of the spectrum energy was bogus.
- The FFT visualization would take a whole input buffer and transform it. That ment frames per second would depend on the size of the buffer and would possibly do very big FFTs which can become CPU intensive. So I took a stab at changing the architecture to make many little subsequent FFTs on the same buffer possible.
- Use log scale because it looks nice :-)

### CODE REVIEW

File: Userland/Applications/SoundPlayer/VisualizationWidget.h [hunk 1]
```diff
@@ -8,15 +8,88 @@
...
+
+        if (m_render_buffer.size() == 0)
+            return;
+
+        size_t buffer_position = (m_frame_count * RefreshTimeMilliseconds) * m_samplerate / 1000;
+        if (buffer_position + m_render_buffer.size() >= m_sample_buffer.size())
+            buffer_position = m_sample_buffer.size() - m_render_buffer.size();
+
+        for (size_t i = 0; i < m_render_buffer.size(); i++)
+            m_render_buffer[i] = m_sample_buffer[i + buffer_position];
```
Comment (kleinesfilmroellchen) [inline Userland/Applications/SoundPlayer/VisualizationWidget.h:59]: Please use AK::TypedTransfer

File: Userland/Applications/SoundPlayer/VisualizationWidget.h [hunk 2]
```diff
@@ -8,15 +8,88 @@
...
+        m_render_buffer.resize(count);
+    }
 
 protected:
-    VisualizationWidget() = default;
+    int m_samplerate;
+    int m_last_buffer_id;
+    size_t m_frame_count;
+    Vector<double> m_sample_buffer;
+    Vector<double> m_render_buffer;
```
Comment (kleinesfilmroellchen) [inline Userland/Applications/SoundPlayer/VisualizationWidget.h:82]: Once again, FixedArray :^)

File: Userland/Applications/SoundPlayer/VisualizationWidget.h [hunk 3]
```diff
@@ -8,15 +8,88 @@
...
+
+    static constexpr size_t RefreshTimeMilliseconds = 30;
```
Comment (kleinesfilmroellchen) [inline Userland/Applications/SoundPlayer/VisualizationWidget.h:84]: As per the code style, this should be SCREAMING_SNAKE_CASE.

File: Userland/Applications/SoundPlayer/BarsVisualizationWidget.h [hunk 4]
```diff
@@ -23,8 +23,13 @@ class BarsVisualizationWidget final : public VisualizationWidget {
     void render(GUI::PaintEvent&, Vector<double> const&) override;
     void context_menu_event(GUI::ContextMenuEvent& event) override;
 
+    static constexpr size_t FFTSize = 256;
+    static constexpr size_t BinGroupCount = 64;
+    static constexpr size_t BinsPerGroup = (FFTSize / 2) / BinGroupCount;
```
Comment (kleinesfilmroellchen) [inline Userland/Applications/SoundPlayer/BarsVisualizationWidget.h:28]: As per the code style, please SCREAMING_SNAKE_CASE here.

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]
Sorry for taking so long, I was busy :(.

Looks great, I'll have a test later this evening!

**Comment by kleinesfilmroellchen** [comment]
Also, please fix your commit message in 1f60f26e2fe4b2f97205624b0d18e5722d2dd2b1

**Comment by kleinesfilmroellchen** [comment]
Just tried the new visualizer and it's definitely much better than the old one, especially for samples. There we won't need any changes. However, there are a couple of debatable technicalities with the new spectral analyzer:
1. The window is now so short that the bars mostly frantically flip about. The best course of action here is probably to just make the window a bit larger again.
2. The windows are used non-overlappingly. This actually makes the spectral analysis even less smooth and may fix (1) so that we don't need shorter windows. I know from writing spectral analysis myself that overlapping windows give much better visuals because two consecutive frames share (up to) half their audio data. This is also how continuous FFT (or short-time fourier transform STFT) is done in most tools, even if it's not used for convolution filters. See https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.stft.html#scipy.signal.stft for an excellent example.

I'd like the frequency adjustment to be the default. It just makes much more sense to boost the higher frequencies by default.

The visualization is currently ahead of the audio, but I'm guessing that's unavoidable with the current time-course audio playback. When I rearchitect audio streaming, this issue will probably be solvable much more easily so I don't expect it to be fixed.

And finally, there seems to be a bug where playback of any file (both FLAC and WAV) stops quite some time before the file has actually finished playing, without any error or anything. This only happens to SoundPlayer (not aplay) and doesn't happen on master so it might be related.

**Comment by standardexe** [comment]
> 

Hey Filmroellchen, unfortunately I don't have a lot of free time at the moment. Thanks for taking over the MP3 decoder PR.

**Comment by kleinesfilmroellchen** [comment]
No problem, I think I'll see this one through as well

**Comment by kleinesfilmroellchen** [comment]
Continued in #12868

**Pull request closed without merging**